### PR TITLE
[WIP] Amphibious vehicles

### PR DESCRIPTION
**Pull request by atlaua**
As the title says, I'm working on making amphibious vehicles possible. There's no code for that yet, though; so far I've only done a minor cleanup of the boat code that I'd love to get some feedback on.

### CODE REVIEW

**Comment by atlaua** [comment]
I've implemented support for amphibious vehicles now; it seems to work fine, but some more testing needs to be done, since this is my first excursion into the depths of the vehicle code.

Also, there's one comfort feature I haven't implemented yet: Currently, all vehicles, including boats, get "is_floating=false" during initialization, which means boats have to be manually set to 'sea' mode by the user after loading a save.

I've tried to recognize boats automatically by adding a `g->m.vehicle_wheel_traction( *this ) < 0` check to `vehicle:refresh()`, but this invariably causes a segfault during save loading that I haven't quite figured out yet. It appears to occur in the `'const auto &tr = ter( pp ).obj();' line in `map::vehicle_wheel_traction()`.

**Comment by BevapDin** [comment]
> I've tried to recognize boats automatically by adding a  g->m.vehicle_wheel_traction( *this ) < 0  check [...] but this invariably causes a segfault during save loading

Most likely because the map has not yet been loaded.

**Comment by atlaua** [comment]
> Most likely because the map has not yet been loaded.

Yes, I assume that's the problem. How do I avoid this, though? I
suppose I could rework vehicle_wheel_traction() to bail out gracefully
in this case.

**Comment by BevapDin** [comment]
Vehicles should dynamically (when needed) detect whether they are floating, instead of relying on a member variable. Instead of having a `is_floating` variable, add a `is_floating()` function.

**Comment by atlaua** [comment]
> Vehicles should dynamically (when needed) detect whether they are
> floating, instead of relying on a member variable. Instead of having
> a `is_floating` variable, add a `is_floating()` function.
> 

I've considered this, but decided against it for now. An
'is_floating()' function would have to loop over all wheels and boat
boards, and by my count, it would be called at least three times for
every single vehicle movement. That seems a bit excessive given that
the job can be easily and much more efficiently done with a variable.

**Comment by kevingranade** [comment]
Even if the internal state that's being checked is a variable, that
variable should be private and callers should check it with a method call.

**Comment by atlaua** [comment]
> Even if the internal state that's being checked is a variable, that
> variable should be private and callers should check it with a method
> call.

Okay, I can change that. I'd just thought that we didn't care too much
about strict encapsulation here since everything's one compilation unit
anyway. There are accesses to other classes' internal variables all
over the place, I haven't seen many accessor functions.

**Comment by atlaua** [comment]
Looking at the code again, making 'is_floating' private with an
accessor doesn't look like a very good idea. I need to set it from
map::vehact(). Also, all the other bools in the vehicle class (other
than three cache invalidation flags), some of them quite comparable in
function to 'is_floating', are public, too.

**Comment by Coolthulhu** [comment]
> An 'is_floating()' function would have to loop over all wheels and boat boards, and by my count, it would be called at least three times for every single vehicle movement.

Requiring an UI operation to switch land/float mode is unacceptable. It clearly shows that the implementation must be wrong or incomplete. Player must not be required to manually select the mode.

You will have to set `is_floating` on the move, in a similar way that `falling` is set. Like this:
* When the vehicle is about to move, check if it could move as a land vehicle
* If it can move as land vehicle, set `is_floating = false`, then move it as a land vehicle
* If it would sink as a land vehicle, check if it is also a boat
  * If it is not a boat, sink it
  * If it is a boat, set `is_floating = true` and move it as a boat

However if `is_floating` is needed earlier, it may require a more significant redesign.

For purposes of UI and the like, it is tolerable (not good, though) if it displays the last state. It may be worth saving this state to savegame json, to avoid potential weirdness on load.

**Comment by atlaua** [comment]
Thanks for the feedback! I don't like the manual mode toggling either,
and I've already lined up two changes to vehicle::thrust() that should
set the mode correctly in all cases. I'll push them as soon as I've
done some more testing.

Currently, I'm only having the code consider a switch to land mode if
it can't move at all in sea mode. I suspect your suggestion would make
for cleaner code without a considerable performance cost, so I'll look
into that.

Serializing the mode may indeed be a good idea; I've already run into
some weirdness when dialing a boat's target speed to down and then
saving and reloading before it has a chance to slow down.

**Comment by atlaua** [comment]
I bit the bullet and made is_floating a function. Everything seems to work fine now, but I need to test and review it more thoroughly.

File: src/vehicle.cpp [hunk 1]
```diff
@@ -5102,7 +5103,9 @@ void vehicle::refresh()
         if( vpi.has_flag( "CAMERA" ) ) {
             camera_epower += vpi.epower;
         }
-        if( vpi.has_flag( VPFLAG_FLOATS ) ) {
+        // The floating cache shouldn't include broken parts;
```
Comment (Coolthulhu) [inline src/vehicle.cpp]: `refresh` is called only on part addition/removal and similar. `floating` must necessarily contain both broken and unbroken parts.

File: src/vehicle.cpp [hunk 2]
```diff
@@ -5632,9 +5635,15 @@ int vehicle::damage_direct( int p, int dmg, damage_type type )
...
-        remove_part( p );
+    }
+
+    if( parts[p].is_broken() ) {
+        if( part_flag(p, "UNMOUNT_ON_DAMAGE") ) {
+            g->m.spawn_item( global_part_pos3( p ), part_info( p ).item, 1, 0, calendar::turn );
+            remove_part( p );
+        }
+
+        refresh();
```
Comment (Coolthulhu) [inline src/vehicle.cpp]: Forcing `refresh` here to keep that weird thing with `floating` not having broken parts sounds really wrong.
Instead make a `can_float` or something like that, which checks if there is a non-broken floating part or something like that.

File: src/vehicle.cpp [hunk 3]
```diff
@@ -6466,3 +6475,24 @@ void vehicle::calc_mass_center( bool use_precalc ) const
...
+    }
+
+    // Pure boat
+    if ( !valid_wheel_config( false ) ) {
+        return true;
+    }
+
+    // Otherwise, the vehicle is amphibious. We consider it
+    // to be floating if >2/3 of the wheels are submerged
+    if ( g->m.vehicle_wheel_traction( *this ) < 0 ) {
```
Comment (Coolthulhu) [inline src/vehicle.cpp]: Negative wheel traction is a hack to sink vehicles. It isn't totally wrong to use it here, but it's not right either.

File: src/vehicle.cpp [hunk 4]
```diff
@@ -6466,3 +6475,24 @@ void vehicle::calc_mass_center( bool use_precalc ) const
         mass_center_no_precalc_dirty = false;
     }
 }
+
+bool vehicle::is_floating() const
```
Comment (Coolthulhu) [inline src/vehicle.cpp]: If you're worried about this being called too often and becoming slow, back it up with a cached variable, just like mass is cached.

**Comment by BorkBorkGoesTheCode** [comment]
Is this dead @atlaua ?

**Comment by atlaua** [comment]
On Sun, 14 Jan 2018 23:40:56 -0800
BorkBorkGoesTheCode <notifications@github.com> wrote:

> Is this dead @atlaua ?
> 

Yep, sorry. It needs a complete rewrite, for which I unfortunately don't
have time right now.

**Comment by BorkBorkGoesTheCode** [comment]
@atlaua what about now?

**Comment by Leland** [comment]
Closing for the time being

**Pull request closed without merging**